File: app/src/main/java/app/endershrooms/thatcatapp/net/CatService.java
package app.endershrooms.thatcatapp.net;

import app.endershrooms.thatcatapp.model.Breed;
import app.endershrooms.thatcatapp.model.Cat;
import app.endershrooms.thatcatapp.model.Category;
import app.endershrooms.thatcatapp.model.Favorite;
import app.endershrooms.thatcatapp.model.RequestResult;
import app.endershrooms.thatcatapp.model.Vote;
import app.endershrooms.thatcatapp.model.builders.FavoriteRequest;
import app.endershrooms.thatcatapp.model.builders.VoteRequest;
import io.reactivex.Single;
import java.util.List;
import java.util.Map;
import retrofit2.http.Body;
import retrofit2.http.DELETE;
import retrofit2.http.GET;
import retrofit2.http.POST;
import retrofit2.http.Path;
import retrofit2.http.Query;
import retrofit2.http.QueryMap;

public interface CatService {

  @GET("breeds")
  Single<List<Breed>> getBreeds();

  @GET("breeds")
  Single<List<Breed>> searchBreeds(@Query("q") String query);

  @GET("categories")
  Single<List<Category>> getCategories();

  @GET("votes")
  Single<Vote> getVote(@Query("vote_id") String id);

  @GET("votes")
  Single<List<Vote>> getVotes();

  @POST("votes")
  Single<RequestResult> createVote(@Body VoteRequest vote);

  @DELETE("votes")
  Single<RequestResult> deleteVote(@Query("vote_id") String id);

  @GET("favourites")
  Single<List<Favorite>> getFavorites(@Query("sub_id") String id,
      @Query("limit") int limit,
      @Query("page") int page);

  @GET("favourites/{favourite_id}")
  Single<Favorite> getFavorite(@Query("sub_id") String id,
      @Path("favourite_id") String favoriteId);

  @DELETE("favourites/{favourite_id}")
  Single<RequestResult> deleteFavorite(@Query("sub_id") String id,
      @Path("favourite_id") String favoriteId);

  @POST("favourites")
  Single<RequestResult> createFavorite(@Body FavoriteRequest favorite);

  @GET("images/search")
  Single<List<Cat>> getImages(@QueryMap Map<String, String> queries);

  @GET("images/{image_id}")
  Single<RequestResult> getImage(@Path("image_id") String id);
}

File: app/src/main/java/app/endershrooms/thatcatapp/net/CatService.kt
package app.endershrooms.thatcatapp.net

import app.endershrooms.thatcatapp.model.*
import app.endershrooms.thatcatapp.model.builders.FavoriteRequest
import app.endershrooms.thatcatapp.model.builders.VoteRequest
import io.reactivex.Single
import retrofit2.http.*

interface CatService {

  @GET("breeds")
  fun getBreeds(): Single<List<Breed>>

  @GET("breeds")
  fun searchBreeds(@Query("q") query: String): Single<List<Breed>>

  @GET("categories")
  fun getCategories(): Single<List<Category>>

  @GET("votes")
  fun getVote(@Query("vote_id") id: String): Single<Vote>

  @GET("votes")
  fun getVotes(): Single<List<Vote>>

  @POST("votes")
  fun createVote(@Body vote: VoteRequest): Single<RequestResult>

  @DELETE("votes")
  fun deleteVote(@Query("vote_id") id: String): Single<RequestResult>

  @GET("favourites")
  fun getFavorites(@Query("sub_id") id: String,
      @Query("limit") limit: Int,
      @Query("page") page: Int): Single<List<Favorite>>

  @GET("favourites/{favourite_id}")
  fun getFavorite(@Query("sub_id") id: String,
      @Path("favourite_id") favoriteId: String): Single<Favorite>

  @DELETE("favourites/{favourite_id}")
  fun deleteFavorite(@Query("sub_id") id: String,
      @Path("favourite_id") favoriteId: String): Single<RequestResult>

  @POST("favourites")
  fun createFavorite(@Body favorite: FavoriteRequest): Single<RequestResult>

  @GET("images/search")
  fun getImages(@QueryMap queries: Map<String, String>):  Single<List<Cat>>

  @GET("images/{image_id}")
  fun getImage(@Path("image_id") id: String)
}